src/clojure_mauth_client/middleware.clj

```diff
@@ -20,10 +20,9 @@
                                 :else (slurp body))
           headers (-> (:headers request)
                       downcase-header-keys)
-          version (get headers "mauth-version")
-          [mauth-time mauth-auth mauth-version] (if (= version "v2")
-                                    [(get headers "mcc-time") (get headers "mcc-authentication") "v2"]
-                                    [(get headers "x-mws-time") (get headers "x-mws-authentication") "v1"])
+          [mauth-time mauth-auth mauth-version] (if (every? headers ["mcc-time" "mcc-authentication"])
```

Comment:
@colinphill-mdsol I disagree. If they ever develop a V3 we need to take a look at how they did it in the past if we want to try to anticipate any changes, and the previous behavior was keeping backwards compatibility. They didn't use the same headers for V2, instead created new ones. So probably that's what's going to happen if a new version is developed. 

If they introduce a breaking change, i.e a V3 using the same header, which seems unlikely. Then we will need to adapt to the new changes and that will most likely require changing the code due to the adoption of another approach different from the token required in v2.

No matter what we do, we cannot anticipate those changes and we shouldn’t  try to. We need to work with the problem at hand. And the solution proposed by @abkumbhar-mdsol  solves that problem.

Reply:
@colinphill-mdsol - I am agree with @omendozarivera-mdsol . I have added else condition to throw exception if no mauth headers are sent in request.

Reply:
If you don't add the mauth headers the response has to be Unauthorized, not an exception. Let it fail.

Reply:
> They didn't use the same headers for V2, instead created new ones. So probably that's what's going to happen if a new version is developed.
No, I don't think that's correct. There would be no reason to add "MWSV2" to the beginning of that header if it weren't meant to be the version declaration moving forward.

In any case, if that string is anything other than "MWSV2", then we have received an unexpected value and should abort. We must not make any assumptions about how a future V3 is supposed to work. Continuing to follow V2 semantics could very likely result in a security problem. If there is any sign that we have something other than the versions we support – V1 and V2 – then we cannot proceed with authentication.

Reply:
It's very common for a standard to neglect versioning in V1, and solve that problem in V2 by just moving to a new field which includes explicit versioning, then continue using that new field indefinitely. Something similar happened in OpenAPI and, I believe, Kubernetes.

Reply:
I've looked again at [the MAuth docs](https://learn.mdsol.com/display/CA/MAuth+Protocol+V2+Specification), and they explain their strategy for versioning. We were all incorrect (emphasis mine):

> For mcc-authentication:
> - "MWSV2" is a string literal indicating the version of the protocol.
> - A space (" ") which separates the protocol version identifier from the app UUID + signature.
> - <app_uuid> the request signer's MAuth app UUID.
> - A colon (":") which separates the app UUID from the digital signature.
> - <digital_signature_for_request> is described below.
> - A semi-colon (";") to indicate the end of the digital signature (since semi-colons are not present in base64-encoding).  **This is present so that in the future, the next version of the protocol could be specified on the same header after this semi-colon, e.g. MWSV2 <app_uuid>:<digital_signature_for_request>; MWSV3 <the_mwsv3_signature>;**

Our parsing needs here are a little more complex than any of us expected. It is complex enough that I think we need to hit the brakes and rethink our approach – any mistake is a potential security problem, so we need to get this right.

There is a project for [mauth-jvm-clients](https://github.com/mdsol/mauth-jvm-clients) which includes both Java and Scala implementations, as well as shared code for both. I propose the following:

1. We revert V2 support in this library. We did not implement it correctly, and a correct implementation would be a significant new effort.
2. As a medium-term tech debt effort, we write a new Clojure MAuth client which is a thin wrapper around the Java implementation in [mauth-jvm-clients](https://github.com/mdsol/mauth-jvm-clients). This will dramatically reduce the risk of making a mistake with security implications, and it will dramatically reduce the amount of work we need to do to maintain the library.
  a. We should also consider contributing it as a module within that same repository, since the Scala implementation is already there. This will also make it easier to benefit from their build process, which seems to include [a standardized test suite](https://github.com/mdsol/mauth-protocol-test-suite).
3. When the new implementation is ready for use, we archive this repository.

Reply:
I would like to review the docs and your proposal before our call to add value to our discussion.

Btw, I agree with Colin on the use of every? I learned a new way to use it. 👍

Reply:
As discussed, we will continue to iterate on this PR, rather than reverting V2 support as I suggested above. When I made that suggestion, I had overlooked the fact that we already released a first pass at V2 support. We should fix the known issues with that first pass just in case someone else is using this library.

We have discussed and reached consensus on migrating to a thin wrapper around the Java client, but this is not something we're doing immediately.

I stand by my initial feedback about checking the version declared in the `mcc-authentication` header, but because we no longer plan to maintain this library indefinitely, I would be fine with a simpler solution than my mockup above. I've left a new comment to discuss that issue on its own.